pkg/updatestatus/nodeinformer.go

```diff
@@ -0,0 +1,390 @@
+package updatestatus
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"gopkg.in/yaml.v3"
+
+	corev1 "k8s.io/api/core/v1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	kubeinformers "k8s.io/client-go/informers"
+	corelistersv1 "k8s.io/client-go/listers/core/v1"
+	"k8s.io/klog/v2"
+
+	machineconfigv1 "github.com/openshift/api/machineconfiguration/v1"
+	configv1client "github.com/openshift/client-go/config/clientset/versioned"
+	machineconfiginformers "github.com/openshift/client-go/machineconfiguration/informers/externalversions"
+	machineconfigv1listers "github.com/openshift/client-go/machineconfiguration/listers/machineconfiguration/v1"
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+
+	"github.com/openshift/cluster-version-operator/pkg/updatestatus/mco"
+)
+
+// nodeInformerController is the controller that monitors health of the node resources
+// and produces insights for node update.
+type nodeInformerController struct {
+	configClient       configv1client.Interface
+	machineConfigs     machineconfigv1listers.MachineConfigLister
+	machineConfigPools machineconfigv1listers.MachineConfigPoolLister
+	nodes              corelistersv1.NodeLister
+	recorder           events.Recorder
+
+	// sendInsight should be called to send produced insights to the update status controller
+	sendInsight sendInsightFn
+
+	// now is a function that returns the current time, used for testing
+	now func() metav1.Time
+}
+
+func newNodeInformerController(
+	configClient configv1client.Interface,
+	coreInformers kubeinformers.SharedInformerFactory,
+	machineConfigInformers machineconfiginformers.SharedInformerFactory,
+	recorder events.Recorder,
+	sendInsight sendInsightFn,
+) factory.Controller {
+	cpiRecorder := recorder.WithComponentSuffix("node-informer")
+
+	c := &nodeInformerController{
+		configClient:       configClient,
+		machineConfigs:     machineConfigInformers.Machineconfiguration().V1().MachineConfigs().Lister(),
+		machineConfigPools: machineConfigInformers.Machineconfiguration().V1().MachineConfigPools().Lister(),
+		nodes:              coreInformers.Core().V1().Nodes().Lister(),
+		recorder:           cpiRecorder,
+		sendInsight:        sendInsight,
+
+		now: metav1.Now,
+	}
+
+	nodeInformer := coreInformers.Core().V1().Nodes().Informer()
+
+	controller := factory.New().
+		// call sync on node changes
+		WithInformersQueueKeysFunc(nodeInformerControllerQueueKeys, nodeInformer).
+		WithSync(c.sync).
+		ToController("NodeInformer", c.recorder)
+
+	return controller
+}
+
+func (c *nodeInformerController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	queueKey := syncCtx.QueueKey()
+
+	t, name, err := parseQueueKey(queueKey)
```

Comment:
`parseQueueKey` is a controlplaneinformer.go method coupled with `configApiQueueKeys`, we should not depend on CPI code. Either we should invent some broad make / parse method pair that handles everything, or we should have separate pairs for each controller (I would prefer the latter)

Reply:
This is a bit borderline - I think the solution I prefer (having separate pairs) can be done here rather than in a followup, but I don't mind if you do it in a followup.